numpy/_core/src/multiarray/array_converter.c

```diff
@@ -0,0 +1,471 @@
+ * where converting inputs to arrays would lose the necessary information.
+ *
+ * The helper thus replaces many asanyarray/asarray calls.
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <structmember.h>
+
+#include "numpy/arrayobject.h"
+#include "arrayobject.h"
+#include "array_converter.h"
+#include "arraywrap.h"
+#include "numpy/arrayscalars.h"
+#include "npy_argparse.h"
+#include "abstractdtypes.h"
+#include "common_dtype.h"
+#include "convert_datatype.h"
+#include "descriptor.h"
+#include "multiarraymodule.h"
+#include "ctors.h"
+
+#include "npy_config.h"
+
+#include "npy_pycompat.h"
+#include "array_assign.h"
+
+#include "common.h"
+#include "get_attr_string.h"
+
+
+
+static PyObject *
+array_converter_new(
+        PyTypeObject *cls, PyObject *args, PyObject *kwds)
+{
+    if (kwds != NULL && PyDict_GET_SIZE(kwds) != 0) {
+        PyErr_SetString(PyExc_TypeError,
+                "Array creation helper doesn't support keywords.");
+        return NULL;
+    }
+
+    Py_ssize_t narrs_ssize_t = (args == NULL) ? 0 : PyTuple_GET_SIZE(args);
+    int narrs = (int)narrs_ssize_t;
+    /* Limit to NPY_MAXARGS for now. */
+    if (narrs_ssize_t > NPY_MAXARGS) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "too many arrays.");
+        return NULL;
+    }
+
+    PyArrayArrayConverterObject *self = PyObject_NewVar(
+            PyArrayArrayConverterObject, cls, narrs);
+    if (self == NULL) {
+        return NULL;
+    }
+    PyObject_InitVar((PyVarObject *)self, &PyArrayArrayConverter_Type, narrs);
+
+    self->narrs = 0;
+    self->flags = 0;
+    self->wrap = NULL;
+    self->wrap_type = NULL;
+
+    if (narrs == 0) {
+        return (PyObject *)self;
+    }
+    self->flags = (NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+
+    creation_item *item = self->items;
+    /* increase self->narrs in loop for cleanup */
+    for (int i = 0; i < narrs; i++, item++) {
+        item->object = PyTuple_GET_ITEM(args, i);
+
+        /* Fast path if input is an array (maybe FromAny should be faster): */
+        if (PyArray_Check(item->object)) {
+            Py_INCREF(item->object);
+            item->array = (PyArrayObject *)item->object;
+            item->scalar_input = 0;
+        }
+        else {
+            item->array = (PyArrayObject *)PyArray_FromAny_int(
+                    item->object, NULL, NULL, 0, 0, 0, NULL,
+                    &item->scalar_input);
+            if (item->array == NULL) {
+                goto fail;
+            }
+        }
+        if (!item->scalar_input) {
+            self->flags &= ~(NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+        }
+        /* At this point, assume cleanup should happen for this item */
+        self->narrs++;
+        Py_INCREF(item->object);
+        item->DType = NPY_DTYPE(PyArray_DESCR(item->array));
+        Py_INCREF(item->DType);
+
+        if (npy_mark_tmp_array_if_pyscalar(
```

Comment:
Could this be `if (item->scalar_input && npy_mark...)`? If so, maybe combine with l.93-95,
```
if (item->scalar_input) {
    if (npy_mark...) {
        ...
    }
    else {
        ...
    }
}
else {  /* non-scalar input */
    self->flags &= ~(NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
}
```

Reply:
Yes, that is correct, and I adopted it, but not as suggested here.  The pretty-code annoyance is that `item->descr` needs to be dealt with, so whatever you do, you have some duplication.

But I agree, it isn't super nice to split this from L93, and maybe easier to grok if it is clear that `npy_mark_tmp_array_if_pyscalar` does indeed require `scalar_input`.